vyos/vyos-1x
# openvpn: T56: check for alternative authentication methods

## Change Summary

My second attempt of https://github.com/vyos/vyos-1x/pull/917, because this PR was reverted due to failing smoketest. I didn't remove these strict checks for the smoketests, so the test failed because it checked if these (removed) checks were working properly.

However, In this new PR, the tls cert-file and key-file config checks are still done but only if no alternative authentication methods are defined. Such as --pks12 or --pkcs11-providers or username/password authentication. I have checked the openvpn source code, and these are currently the alternative methods that can be configured instead of --cert and --key.

This change allows advanced users to configure these settings via "openvpn-option" and not be limited by the tls cert-file and key-file configuration check.

## Types of changes


Only check for TLS cert-file and key-file if no alternative authentication methods are defined. This should not impact any functionality or usability.

## Related Task(s)

* https://phabricator.vyos.net/T56

## Component(s) name

openvpn

## Proposed changes


## How to test


Given a valid `/config/auth/vpn/ca.crt` file, 

This should result in Missing "tls cert-file"
```
del interfaces openvpn vtun1
set interfaces openvpn vtun1 remote-host 127.0.0.1
set interfaces openvpn vtun1 mode client
set interfaces openvpn vtun1 tls ca-cert-file /config/auth/vpn/ca.crt
commit
```

This should pass (--pkcs11-providers):
```
del interfaces openvpn vtun1
set interfaces openvpn vtun1 remote-host 127.0.0.1
set interfaces openvpn vtun1 mode client
set interfaces openvpn vtun1 tls ca-cert-file /config/auth/vpn/ca.crt
set interfaces openvpn vtun1 openvpn-option "--pkcs11-providers /usr/lib/libeToken.so"
commit
```

And this should also pass (--pkcs12):

```
del interfaces openvpn vtun1
set interfaces openvpn vtun1 remote-host 127.0.0.1
set interfaces openvpn vtun1 mode client
set interfaces openvpn vtun1 tls ca-cert-file /config/auth/vpn/ca.crt
set interfaces openvpn vtun1 openvpn-option "--pkcs12 /config/auth/vpn/client.p12"
commit
```

And this should also pass (username/password auth only):

```
del interfaces openvpn vtun1
set interfaces openvpn vtun1 remote-host 127.0.0.1
set interfaces openvpn vtun1 mode client
set interfaces openvpn vtun1 tls ca-cert-file /config/auth/vpn/ca.crt
set interfaces openvpn vtun1 authentication username user
set interfaces openvpn vtun1 authentication password pass
commit
```

## Checklist:
